src/ini.cr

```diff
@@ -1,22 +1,52 @@
 class INI
+  # Exception thrown on an INI parse error.
+  class ParseException < Exception
+    getter line_number : Int32
+    getter column_number : Int32
+
+    def initialize(message, @line_number, @column_number)
+      super "#{message} at #{@line_number}:#{@column_number}"
+    end
+
+    def location
+      {line_number, column_number}
+    end
+  end
+
   # Parses INI-style configuration from the given string.
+  # Raises a `ParseException` on any errors.
   #
   # ```
   # INI.parse("[foo]\na = 1") # => {"foo" => {"a" => "1"}}
   # ```
   def self.parse(str) : Hash(String, Hash(String, String))
     ini = {} of String => Hash(String, String)
-
     section = ""
-    str.each_line do |line|
-      if line =~ /\s*(.*[^\s])\s*=\s*(.*[^\s])/
-        ini[section] ||= {} of String => String if section == ""
-        ini[section][$1] = $2
-      elsif line =~ /\[(.*)\]/
-        section = $1
+
+    str.lines.each_with_index(1) do |oline, lineno|
+      line = oline.strip
```

Comment:
Simply stripping whitespace renders the column indices in error messages meaningless. You'll have to skip and count whitespace to get a valid offset.

Reply:
Yeah, that's what I was going for with `oline` (which is supposed to be **o**riginal **line**), but I realized that's not sufficient. I'll turn skip-and-count into a private helper method, since it's needed in a few places.